torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -1526,7 +1528,16 @@ bool FunctionSignature::parse(
       }
       return false;
     } else if (param.check(obj, overloaded_args, i, &failed_idx)) {
-      dst[i++] = obj;
+      if (param.type_ == ParameterType::DEVICE &&
+          torch::is_symint(py::handle(obj))) {
+        // Cast to c10::SymInt first to guard_int with stack trace
+        // then cast it back to python.
+        auto guarded_int = py::cast<c10::SymInt>(py::handle(obj))
+                               .guard_int(__FILE__, __LINE__);
+        dst[i++] = py::cast(guarded_int).ptr();
```

Comment:
This sure looks like it will result in an excessive decref later (if dst is owning). Can you argue why this has the right lifetime? (cc @albanD)

Reply:
Yeah, the life time is wrong. I'm not very familar to pybind and digged a bit into it: dst doesn't own `py::cast(guarded_int)`. The integer is a local variable so dst will have a dangling pointer to it after we exits the if-statement. The previous diff works probably because gc hasn't cleaned the symint up. Thanks for pointing it out! 

Updated the diff so that when we're actually creating Device from argument, we do the specialization and create a device using integer index.